taichi/backends/metal/api.cpp

```diff
@@ -64,13 +64,14 @@ nsobj_unique_ptr<MTLBlitCommandEncoder> new_blit_command_encoder(
 
 nsobj_unique_ptr<MTLLibrary> new_library_with_source(MTLDevice *device,
                                                      const std::string &source,
+                                                     bool fast_math,
                                                      int msl_version) {
   auto source_str = mac::wrap_string_as_ns_string(source);
 
   id options = clscall("MTLCompileOptions", "alloc");
   options = call(options, "init");
   auto options_cleanup = wrap_as_nsobj_unique_ptr(options);
-  call(options, "setFastMathEnabled:", false);
+  call(options, "setFastMathEnabled:", fast_math);
```

Comment:
Cool! Not sure about how `fast_math` works. IIUC is this the same as specifying [`precision mediump float;`](https://blog.csdn.net/weixin_45360983/article/details/100084990) in OpenGL?

Reply:
They are both for optimizing performance, but take different approaches. From what i can tell, `precision mediump float` reduces the bits to represent float. Metal also has a similar concept of `half` (16bits). Fast math, on the other hand, reduces the instructions in the computation to produce an approximated result.